# HOW TO USE ISO 14053:2021(E) MFCA C# Library

The results of this project showcase an innovative C#-developed library crafted to empower users in understanding and applying the principles of ISO 14053:2021(E). To enhance the accessibility of this intricate standard, I harnessed the versatility of Jupyter Notebook to create engaging, interactive documentation and dynamic code examples. This seamless fusion of theoretical insights and hands-on programming not only makes learning more enjoyable but also equips users with a thorough guide for effective implementation.

Mahler Chou (周敬斐)

### To use it for the first time, you must import the MfcaLib.cs library

In [19]:
#load "MfcaLibs.cs"

### Create a process instance/object to hold data

According to ISO 14053, the term previously known as "Quantity Center" has been updated to "Process" to better align with standard terminology. 

The C# code sample below demonstrates how to declare a Process object variable named `p1`. When adding materials, waste treatment costs, and other related parameters, the variable `p` is used to define the scope of that process. This approach allows for flexibility; if you have multiple processes in the future, you can name them as you wish, such as `a1`, `a2`, and `a3`.

When declaring a Process variable, you must provide the process name, start date, operation time, and expected completion quantity in that order. This basic information will be utilized for generating the final report.

In [20]:
Process p1 = new("Steel plate cutting", "10/1", "(60 days)", 2000);

### TARGET MATERIALS
We assume this process is a steel plate cutting department. The main raw materials input into it are steel plates of various sizes and thicknesses, each with a different unit price and quantity.

| Item Name | Unit Price |Quantity |Cost  |Product Quantity |Loss Quantity |
| ---:      | ---:       | ---:    | ---: | ---:            | ---:         |   
| Steel sheet A|13|1207|15691|896|311|
| Steel sheet B|8|600|4800|223|377 |
| Steel sheet C|30|120|3600|88|32 |
| Steel sheet D|20|602|12040|455|147|
| Steel sheet E|36|180|6480|135|45|


In [21]:
p1.SetTargetMaterial("Steel sheet A", 13, 896, 311);
p1.SetTargetMaterial("Steel sheet B", 8, 223, 377);
p1.SetTargetMaterial("Steel sheet C", 30, 88, 32);
p1.SetTargetMaterial("Steel sheet D", 20, 455, 147);
p1.SetTargetMaterial("Steel sheet E", 36, 135, 45);

### Show material items in the process.
After adding the material items into process object p1, you can call p1.ShowMaterial() print out all the material items in the process. 
> _Please note that the output results will show the material's allocation ratio (Allocation Ratio), which is used to allocate energy costs and system costs._

In [22]:
p1.ShowMaterial()

(TARGET MATERIALS)
Material Name        Unit Price Input      Cost       | Product    Cost       | Loss       Cost       | Loss Ratio 
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ---------- | ----------
Steel sheet A                13       1207      15691 |        896      11648 |        311       4043 |      25.8%
Steel sheet B                 8        600       4800 |        223       1784 |        377       3016 |      62.8%
Steel sheet C                30        120       3600 |         88       2640 |         32        960 |      26.7%
Steel sheet D                20        602      12040 |        455       9100 |        147       2940 |      24.4%
Steel sheet E                36        180       6480 |        135       4860 |         45       1620 |      25.0%
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ---------- | 
Subtotal of MTL.                      2709      42611 |       1797    

> _In the production process, there are certain materials that, while not classified as primary raw materials, still play a vital role and are included in the material input list. Take lubricating oil, for example; it’s essential for ensuring that machines operate smoothly when cutting steel plates. Although lubricating oil is considered a non-primary material with a relatively minor impact on overall costs, its importance cannot be overlooked. Consequently, it is excluded from the calculation of the 'positive/negative product ratio,' a key metric that significantly influences the assessment of energy and system costs. This thoughtful approach underscores our commitment to accuracy and efficiency in production._

In [23]:
p1.SetTargetMaterial("Machine oil", 2, 0, 28, false);
p1.ShowMaterial()

(TARGET MATERIALS)
Material Name        Unit Price Input      Cost       | Product    Cost       | Loss       Cost       | Loss Ratio 
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ---------- | ----------
Steel sheet A                13       1207      15691 |        896      11648 |        311       4043 |      25.8%
Steel sheet B                 8        600       4800 |        223       1784 |        377       3016 |      62.8%
Steel sheet C                30        120       3600 |         88       2640 |         32        960 |      26.7%
Steel sheet D                20        602      12040 |        455       9100 |        147       2940 |      24.4%
Steel sheet E                36        180       6480 |        135       4860 |         45       1620 |      25.0%
* Machine oil                 2         28         56 |          0          0 |         28         56 |     100.0%
-------------------- ---------- ---------- ---------- | ----

### WASTE MANAGEMENT COST
Use `SetWasteMgmtCost(item, cost)` to add waste item cost to the process. Then use `ShowWasteMgmt()` print out the waste items list.

In [24]:
p1.SetWasteMgmtCost("Waste steel scrapts", 365);
p1.ShowWasteMgmt();

(WASTE MANAGEMENT)
Item Name                                             |                       |            Cost      
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------
Waste steel scrapts                                   |                       |            365       
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------
Subtotal                                              |                       |            365       
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------


### ENERGY COST
Use `SetEneryCost(item, cost)` to add the energy item. Calling `ShowEnergyCost()` to print the energy cost items in the process.

> _Please note that the energy cost allocation for this run is calculated based on the material allocation (product/loss) ratio._

In [25]:
p1.SetEnergyCost("Electricity", 1660);
p1.ShowEnergyCost();

(ENERGY COSTS)
Item Name            Unit Price Quantity   Cost       | Ratio(+)   Cost       | Ratio(-)   Cost      
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------
Electricity                                      1660 |      66.3%       1101 |      33.7%        559
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------
SUBTOTAL                                        1,660 |                 1,101 |                   559
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------


# SYSTEM COST

Additional costs such as equipment depreciation, machine operating costs, labor, and so on, are all included in the system cost. To add system costs, please use the `SetSystemCost(item, cost)` function. To view the system cost items within process p1, use the `ShowSystemCost()` function.

> _Please note that the system cost allocation for this run is also calculated based on the material allocation (product/loss) ratio._

In [26]:
p1.SetSystemCost("Labour", 45000);
p1.SetSystemCost("Depreciation", 58000);
p1.SetSystemCost("Other", 15000);
p1.ShowSystemCost();

(SYSTEM COSTS)
Item Name            Unit Price Quantity   Cost       | Ratio(+)   Cost       | Ratio(-)   Cost      
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------
Labour                                          45000 |      66.3%      29835 |      33.7%      15165
Depreciation                                    58000 |      66.3%      38454 |      33.7%      19546
Other                                           15000 |      66.3%       9945 |      33.7%       5055
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------
SUBTOTAL                                      118,000 |      66.3%     78,234 |      33.7%     39,766
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ----------


## MFCA SUMMARY SHEET

In this compelling scenario, process `p1` focuses on a steel-cutting operation. Following the completion of data entry, you can easily invoke the Report method to generate a detailed calculation summary table. This table and the accompanying scenario are meticulously aligned with the standards set by ISO 14053:2021(E), `Table A.5`, ensuring the accuracy and reliability of the results.

The calculation summary reveals a significant financial loss of approximately `$55,325` in this process, accounting for an astonishing `32.8%` of the total cost. This highlights a remarkable opportunity for improvement at this process station—particularly alarming is the scrap ratio for 'Steel Sheet B,' which stands at a staggering 62.8%.

With the provided code and usage instructions, you can swiftly calculate losses across each process station. Moreover, these insights can be seamlessly integrated with work orders, routing, and BOM tables, or leveraged through automated procedures for consistent tracking and auditing in your ERP/MES systems. This vividly embodies the essence of Material Flow Cost Accounting (MFCA), which seeks to uncover hidden cost losses through the lens of material flow. Embracing this approach can pave the way for enhanced efficiency and profitability.

In [27]:
p1.Report();

Selected production process: Steel plate cutting
Production period: 10/1 - (60 days)
Planned production volumn: 2000
(TARGET MATERIALS)
Material Name        Unit Price Input      Cost       | Product    Cost       | Loss       Cost       | Loss Ratio 
-------------------- ---------- ---------- ---------- | ---------- ---------- | ---------- ---------- | ----------
Steel sheet A                13       1207      15691 |        896      11648 |        311       4043 |      25.8%
Steel sheet B                 8        600       4800 |        223       1784 |        377       3016 |      62.8%
Steel sheet C                30        120       3600 |         88       2640 |         32        960 |      26.7%
Steel sheet D                20        602      12040 |        455       9100 |        147       2940 |      24.4%
Steel sheet E                36        180       6480 |        135       4860 |         45       1620 |      25.0%
* Machine oil                 2         28         56 |   